deep_qa/layers/attention/weighted_sum.py

```diff
@@ -71,39 +71,51 @@ def compute_output_shape(self, input_shapes):
 
     @overrides
     def call(self, inputs, mask=None):
+        # pylint: disable=redefined-variable-type
         matrix, attention_vector = inputs
-        matrix_shape = K.int_shape(matrix)
-        matrix = self._expand_matrix_if_necessary(matrix, matrix_shape[:-1], attention_vector)
+        matrix = self._expand_matrix_if_necessary(matrix, attention_vector, K.ndim(matrix) - 1)
```

Comment:
This is a fancy weighted sum =). We actually need this in BiDAF, because we're doing a weighted sum of the question, for each passage word.  The inputs [here](https://github.com/allenai/deep_qa/blob/67ce4ebc172ac36d329840b5641806b565982225/deep_qa/models/reading_comprehension/bidirectional_attention.py#L128) are of shape `(batch_size, question_length, encoding_dim)` (for the "matrix"), and `(batch_size, passage_length, question_length)`, for the "attention vector".  We want the result to be `(batch_size, passage_length, encoding_dim)`, summing over the question length.  To do this, we need to expand the "matrix" first.

See also this docstring: 
https://github.com/allenai/deep_qa/blob/67ce4ebc172ac36d329840b5641806b565982225/deep_qa/layers/attention/weighted_sum.py#L37-L53.